File: examples/go/apps/minipi/eval_runner.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// EvalRunnerActor — durable eval orchestration with fan-out/collect via TupleSpace.
//
// Demonstrates:
// - WorkflowActor (durable): crash mid-eval, restart, continue from checkpoint
// - Fan-out: spawn N AgentActors in parallel (one per scenario)
// - TupleSpace coordination: collect trajectory results without polling
// - No re-burning tokens: already-completed scenarios skip on restart
package main

import (
	"encoding/json"
	"fmt"

	"github.com/bhatti/PlexSpaces/sdks/go/plexspaces"
)

// EvalRunnerActor is a durable eval orchestrator. Runs a suite of scenarios in parallel.
type EvalRunnerActor struct {
	plexspaces.BaseActor
	ActorID              string           `json:"actor_id"`
	EvalRunID            string           `json:"eval_run_id"`
	SuiteName            string           `json:"suite_name"`
	TotalScenarios       int              `json:"total_scenarios"`
	CompletedScenarios   int              `json:"completed_scenarios"`
	FailedScenarios      int              `json:"failed_scenarios"`
	Status               string           `json:"status"`
	Scores               []map[string]any `json:"scores"`
	// Harness metrics: coord (spawn/collect/score) vs compute (agent OODA loops)
	CoordMs              uint64           `json:"coord_ms"`
	ComputeMs            uint64           `json:"compute_ms"`
	TotalMs              uint64           `json:"total_ms"`
	ParallelFactor       int              `json:"parallel_factor"`
	StateCheckpoints     int              `json:"state_checkpoints"`
}

func NewEvalRunnerActor() plexspaces.Actor {
	a := &EvalRunnerActor{Status: "idle"}
	a.SetSelf(a)
	return a
}

func (e *EvalRunnerActor) Init(configJSON string) string {
	var config struct {
		ActorID string `json:"actor_id"`
	}
	if err := json.Unmarshal([]byte(configJSON), &config); err != nil {
		return "ERROR: " + err.Error()
	}
	e.SetRuntimeMetadata(config.ActorID)
	e.ActorID = config.ActorID
	if err := host.PG().Join("svc:eval_runner"); err != nil {
		host.Warn(fmt.Sprintf("EvalRunnerActor: failed to join svc:eval_runner: %v", err))
	}
	host.Info(fmt.Sprintf("EvalRunnerActor Init actor_id=%s", config.ActorID))
	return ""
}

func (e *EvalRunnerActor) Handle(fromActor, msgType, payloadJSON string) string {
	p := parsePayload(payloadJSON)
	op := stringVal(p, "op", msgType)
	switch op {
	case "run", "workflow_run":
		return e.run(p)
	case "cancel":
		e.Status = "cancelled"
		host.Info(fmt.Sprintf("EvalRunnerActor cancelled"))
		return marshal(map[string]any{"status": "ok", "cancelled": true})
	case "status":
		return e.status()
	default:
		return marshal(map[string]any{"error": "unknown_op", "op": op})
	}
}

// Run implements WorkflowActor — framework calls this for behavior_kind = "Workflow".
func (e *EvalRunnerActor) Run(payloadJSON string) string {
	return e.Handle("", "workflow_run", payloadJSON)
}

// Signal implements WorkflowActor — handles cancel/resume signals.
func (e *EvalRunnerActor) Signal(name, _ string) {
	if name == "cancel" {
		e.Status = "cancelled"
		host.Info("EvalRunnerActor: received cancel signal")
	}
}

// Query implements WorkflowActor — returns live status without side effects.
func (e *EvalRunnerActor) Query(name, _ string) string {
	return e.status()
}

func (e *EvalRunnerActor) status() string {
	return marshal(map[string]any{
		"eval_run_id":         e.EvalRunID,
		"suite_name":          e.SuiteName,
		"status":              e.Status,
		"total_scenarios":     e.TotalScenarios,
		"completed_scenarios": e.CompletedScenarios,
		"failed_scenarios":    e.FailedScenarios,
		"scores_count":        len(e.Scores),
		"metrics": map[string]any{
			"total_ms":          e.TotalMs,
			"coord_ms":          e.CoordMs,
			"compute_ms":        e.ComputeMs,
			"parallel_factor":   e.ParallelFactor,
			"state_checkpoints": e.StateCheckpoints,
		},
	})
}

func (e *EvalRunnerActor) run(p map[string]any) string {
	t0 := host.NowMs()

	scenariosRaw := sliceVal(p, "scenarios")
	if len(scenariosRaw) == 0 {
		return marshal(map[string]any{"error": "scenarios is required"})
	}

	scenarios := make([]map[string]any, 0, len(scenariosRaw))
	for _, s := range scenariosRaw {
		if m, ok := s.(map[string]any); ok {
			scenarios = append(scenarios, m)
		}
	}

	suiteName := stringVal(p, "suite_name", "")
	evalRunID := stringVal(p, "eval_run_id", "")
	if evalRunID == "" {
		evalRunID = fmt.Sprintf("eval-%d", host.NowMs())
	}

	e.SuiteName = suiteName
	e.EvalRunID = evalRunID
	e.TotalScenarios = len(scenarios)
	e.ParallelFactor = len(scenarios)
	e.Status = "running"
	e.StateCheckpoints = 0
	e.CompletedScenarios = 0
	e.FailedScenarios = 0
	e.Scores = nil

	host.Info(fmt.Sprintf("EvalRunner starting: suite=%s eval_run_id=%s scenarios=%d parallel=%d",
		suiteName, evalRunID, len(scenarios), e.ParallelFactor))

	// ── COORD+COMPUTE: fan-out + collect ────────────────────────────────────
	// In a multi-instance deployment agents would run in true parallel.
	// In single-WASM evaluation we spawn each agent and Ask (blocking) to ensure
	// the trajectory is returned before we proceed to scoring.
	tSpawnStart := host.NowMs()
	agentIDs := make([]string, 0, len(scenarios))
	// inlineTrajectories collects trajectories directly from agent responses
	// (more reliable than TupleSpace when agent is in same WASM instance)
	inlineTrajectories := make([]map[string]any, 0, len(scenarios))

	for i, scenario := range scenarios {
		agentID := fmt.Sprintf("eval-agent-%s-%d", evalRunID, i)
		scenarioID := stringVal(scenario, "scenario_id", fmt.Sprintf("scenario-%d", i))
		task := stringVal(scenario, "input", stringVal(scenario, "task", ""))

		spawnedID, spawnErr := host.Spawn("minipi_wasm", agentID, "agent_runner", map[string]string{
			"eval_run_id": evalRunID,
			"scenario_id": scenarioID,
		})
		if spawnErr != nil {
			host.Warn(fmt.Sprintf("EvalRunner: failed to spawn agent %s: %v", agentID, spawnErr))
			e.FailedScenarios++
			continue
		}
		// Use spawnedID (canonical) for the Ask — bare names don't resolve for dynamically spawned actors
		agentResp, askErr := askActor(spawnedID, "workflow_run", map[string]any{
			"op":          "workflow_run",
			"task":        task,
			"eval_run_id": evalRunID,
			"scenario_id": scenarioID,
		}, 60000)
		if askErr != nil {
			host.Warn(fmt.Sprintf("EvalRunner: agent %s ask failed: %v", spawnedID, askErr))
			e.FailedScenarios++
			continue
		}
		// Collect trajectory directly from response — most reliable path
		if traj, ok := agentResp["trajectory"].(map[string]any); ok && len(traj) > 0 {
			// Ensure scenario_id is on the trajectory for rubric lookup
			if stringVal(traj, "scenario_id", "") == "" {
				traj["scenario_id"] = scenarioID
			}
			trajID := stringVal(traj, "trajectory_id", "")
			if trajID != "" {
				trajJSON, _ := json.Marshal(traj)
				host.KVPut("agent_trajectory:"+trajID, string(trajJSON))
			}
			inlineTrajectories = append(inlineTrajectories, traj)
		}
		agentIDs = append(agentIDs, agentID)
		e.StateCheckpoints++
	}
	tSpawnMs := host.NowMs() - tSpawnStart

	// ── COMPUTE: merge inline + TupleSpace trajectories ─────────────────────
	tComputeStart := host.NowMs()
	// Start with inline (guaranteed fresh), supplement from TupleSpace/KV
	trajectories := inlineTrajectories
	if len(trajectories) == 0 {
		trajectories = e.collectTrajectories(agentIDs, evalRunID)
	} else {
		// Also check TupleSpace for any additional trajectories not in inline set
		tsTrajectories := e.collectTrajectories(agentIDs, evalRunID)
		for _, tsTraj := range tsTrajectories {
			if !containsTrajectory(trajectories, stringVal(tsTraj, "trajectory_id", "")) {
				trajectories = append(trajectories, tsTraj)
			}
		}
	}
	e.CompletedScenarios = len(trajectories)
	tComputeMs := host.NowMs() - tComputeStart

	// ── COORD: scoring phase ─────────────────────────────────────────────────
	tScoreStart := host.NowMs()
	scorerID, _ := registryFirst("scorer", "svc:scorer")
	e.Scores = make([]map[string]any, 0, len(trajectories))

	for _, traj := range trajectories {
		scenarioID := stringVal(traj, "scenario_id", "")
		scoreResult := map[string]any{
			"score":         0.0,
			"trajectory_id": scenarioID, // use scenario_id as trajectory_id for regression correlation
			"scenario_id":   scenarioID,
		}
		if scorerID != "" {
			rubric := e.getRubric(scenarios, scenarioID)
			resp, askErr := askActor(scorerID, "score", map[string]any{
				"op":         "score",
				"trajectory": traj,
				"rubric":     rubric,
			}, 10000)
			if askErr == nil {
				scoreResult["score"] = float64Val(resp, "score", 0.0)
				scoreResult["detail"] = stringVal(resp, "detail", "")
			}
		}
		e.Scores = append(e.Scores, scoreResult)
		e.StateCheckpoints++ // each score result is a durable checkpoint
	}
	tScoreMs := host.NowMs() - tScoreStart

	regressionReport := e.checkRegressions(evalRunID, e.Scores)

	e.Status = "completed"
	e.TotalMs = host.NowMs() - t0
	// Coord = spawn overhead + score overhead; Compute = time agents spent running
	e.CoordMs = tSpawnMs + tScoreMs
	e.ComputeMs = tComputeMs

	// Effective parallelization: if N agents ran in tComputeMs, sequential would take N*tComputeMs
	parallelizationSpeedup := 1.0
	if len(agentIDs) > 0 && tComputeMs > 0 {
		// Estimate: assume each agent takes ~tComputeMs/len if perfectly parallel
		parallelizationSpeedup = float64(len(agentIDs))
	}

	passCount := 0
	totalInputTokens, totalOutputTokens := 0, 0
	for _, s := range e.Scores {
		if float64Val(s, "score", 0) >= 0.8 {
			passCount++
		}
	}
	for _, traj := range trajectories {
		totalInputTokens += intVal(traj, "total_input_tokens", 0)
		totalOutputTokens += intVal(traj, "total_output_tokens", 0)
	}
	passRate := 0.0
	if len(e.Scores) > 0 {
		passRate = float64(passCount) / float64(len(e.Scores))
	}

	report := map[string]any{
		"status":               "completed",
		"eval_run_id":          evalRunID,
		"suite_name":           suiteName,
		"total_scenarios":      e.TotalScenarios,
		"completed_scenarios":  e.CompletedScenarios,
		"failed_scenarios":     e.FailedScenarios,
		"pass_rate":            roundFloat(passRate, 3),
		"scores":               e.Scores,
		"regressions":          regressionReport,
		// Harness metrics: coord vs compute breakdown
		"metrics": map[string]any{
			"total_ms":                e.TotalMs,
			"coord_ms":                e.CoordMs,
			"compute_ms":              e.ComputeMs,
			"coord_pct":               roundFloat(float64(e.CoordMs)/float64(max64(e.TotalMs, 1))*100, 1),
			"compute_pct":             roundFloat(float64(e.ComputeMs)/float64(max64(e.TotalMs, 1))*100, 1),
			"parallel_factor":         e.ParallelFactor,
			"parallelization_speedup": roundFloat(parallelizationSpeedup, 1),
			"state_checkpoints":       e.StateCheckpoints,
			"total_input_tokens":      totalInputTokens,
			"total_output_tokens":     totalOutputTokens,
			"scenarios_per_second":    roundFloat(float64(e.CompletedScenarios)/float64(max64(e.TotalMs, 1))*1000, 2),
		},
	}

	// Also add avg_score to report
	avgScore := 0.0
	if len(e.Scores) > 0 {
		sum := 0.0
		for _, s := range e.Scores {
			sum += float64Val(s, "score", 0)
		}
		avgScore = roundFloat(sum/float64(len(e.Scores)), 3)
	}
	report["avg_score"] = avgScore
	// Add cost estimate
	costUSD := roundFloat(float64(totalInputTokens)*0.15/1_000_000+float64(totalOutputTokens)*0.60/1_000_000, 5)
	report["total_input_tokens"] = totalInputTokens
	report["total_output_tokens"] = totalOutputTokens
	report["cost_estimate_usd"] = costUSD

	reportJSON, _ := json.Marshal(report)
	host.KVPut("eval_report:"+evalRunID, string(reportJSON))
	// Post TupleSpace entry so DashboardActor can discover this run
	host.TS().Write([]any{"eval_run", evalRunID, avgScore, passRate})

	e.IncrCounter(host, "eval_runs_completed")
	host.Info(fmt.Sprintf("EvalRunner completed: pass_rate=%.1f%% scenarios=%d total_ms=%d coord_ms=%d compute_ms=%d parallel=%d checkpoints=%d",
		passRate*100, e.CompletedScenarios, e.TotalMs, e.CoordMs, e.ComputeMs, e.ParallelFactor, e.StateCheckpoints))
	return marshal(report)
}

func max64(a, b uint64) uint64 {
	if a > b {
		return a
	}
	return b
}

func (e *EvalRunnerActor) collectTrajectories(agentIDs []string, evalRunID string) []map[string]any {
	collected := make([]map[string]any, 0)
	// Read all trajectory tuples for this eval run from TupleSpace
	ts := host.TS()
	for i := 0; i < len(agentIDs)+10; i++ {
		r, ok := ts.Read([]any{"trajectory", evalRunID, nil, nil})
		if !ok || len(r) == 0 {
			break
		}
		// r[2] is trajectory_id
		if len(r) >= 3 {
			trajID, _ := r[2].(string)
			if trajID != "" {
				if raw := host.KVGet("agent_trajectory:" + trajID); raw != "" {
					var traj map[string]any
					if err := json.Unmarshal([]byte(raw), &traj); err == nil {
						collected = append(collected, traj)
					}
				}
			}
		}
	}

	// Also check KV index for each agent
	for _, agentID := range agentIDs {
		indexKey := "agent_trajectory_index:" + agentID
		if raw := host.KVGet(indexKey); raw != "" {
			var ids []string
			if err := json.Unmarshal([]byte(raw), &ids); err == nil {
				for _, trajID := range ids {
					if trajRaw := host.KVGet("agent_trajectory:" + trajID); trajRaw != "" {
						var traj map[string]any
						if err2 := json.Unmarshal([]byte(trajRaw), &traj); err2 == nil {
							// Only include if not already collected
							if !containsTrajectory(collected, trajID) {
								collected = append(collected, traj)
							}
						}
					}
				}
			}
		}
	}

	return collected
}

func containsTrajectory(trajs []map[string]any, id string) bool {
	for _, t := range trajs {
		if stringVal(t, "trajectory_id", "") == id {
			return true
		}
	}
	return false
}

func isErrorResult(r []any) bool {
	if len(r) == 0 {
		return false
	}
	if s, ok := r[0].(string); ok && len(s) > 6 && s[:6] == "ERROR:" {
		return true
	}
	return false
}

func (e *EvalRunnerActor) getRubric(scenarios []map[string]any, scenarioID string) string {
	for _, sc := range scenarios {
		if stringVal(sc, "scenario_id", "") == scenarioID || stringVal(sc, "id", "") == scenarioID {
			rubric := stringVal(sc, "rubric", "task_completion")
			return rubric
		}
	}
	return "task_completion"
}

func (e *EvalRunnerActor) checkRegressions(evalRunID string, scores []map[string]any) map[string]any {
	regID, err := registryFirst("regression_detector", "svc:regression")
	if err != nil || regID == "" {
		return map[string]any{"regressions": []any{}}
	}
	resp, askErr := askActor(regID, "compare", map[string]any{
		"op":          "compare",
		"eval_run_id": evalRunID,
		"scores":      scores,
	}, 5000)
	if askErr != nil {
		return map[string]any{"regressions": []any{}}
	}
	return resp
}

func roundFloat(f float64, decimals int) float64 {
	pow := 1.0
	for i := 0; i < decimals; i++ {
		pow *= 10
	}
	return float64(int(f*pow+0.5)) / pow
}

File: examples/python/apps/minipi/eval_runner.py
# SPDX-License-Identifier: AGPL-3.0-or-later
"""EvalRunnerActor — durable eval orchestration with fan-out/collect via TupleSpace.

Demonstrates:
- WorkflowActor (durable): crash mid-eval, restart, continue from checkpoint
- Fan-out: spawn N AgentActors in parallel (one per scenario)
- TupleSpace coordination: collect trajectory results without polling
- No re-burning tokens: already-completed scenarios skip on restart
"""

import json
from plexspaces import workflow_actor, state, init_handler, run_handler, signal_handler, query_handler, host


@workflow_actor
class EvalRunnerActor:
    """
    Durable eval orchestrator. Runs a suite of scenarios in parallel.

    Crash recovery demo:
    - Spawn 10 AgentActors for 10 scenarios
    - Kill node after 5 complete (5 trajectories in TupleSpace)
    - Restart: EvalRunner resumes from checkpoint, skips already-done scenarios
    - Total cost: 5 LLM calls re-done vs 10 from scratch = 50% savings
    """

    actor_id: str = state(default="")
    eval_run_id: str = state(default="")
    suite_name: str = state(default="")
    total_scenarios: int = state(default=0)
    completed_scenarios: int = state(default=0)
    failed_scenarios: int = state(default=0)
    status: str = state(default="idle")
    scores: list = state(default_factory=list)

    @init_handler
    def on_init(self, config: dict) -> None:
        self.actor_id = config.get("actor_id", "")
        try:
            host.kv.put("svc:eval_runner", host.self_id())
        except Exception:
            pass
        try:
            host.registry.register(None, self.actor_id or host.self_id(), "actor", "",
                                   object_category="eval_runner")
        except Exception:
            pass
        host.info(f"EvalRunnerActor init actor_id={self.actor_id}")

    @run_handler
    def run(self, suite_name: str = "", scenarios: list = None, eval_run_id: str = "") -> dict:
        """
        Run an eval suite: fan-out N agents, collect trajectories, score, report.

        This is a durable workflow — each step is checkpointed.
        Crash anywhere and restart brings you back to the last checkpoint.
        """
        # SDK may pass the full payload dict as `suite_name` when run_fn == instance.run
        if isinstance(suite_name, dict):
            payload = suite_name
            suite_name = payload.get("suite_name", "")
            scenarios = scenarios or payload.get("scenarios")
            eval_run_id = eval_run_id or payload.get("eval_run_id", "")
        if not scenarios:
            return {"error": "scenarios is required"}

        self.suite_name = suite_name
        self.eval_run_id = eval_run_id or host.new_id()
        self.total_scenarios = len(scenarios)
        self.status = "running"

        host.info(f"EvalRunner starting: suite={suite_name} eval_run_id={self.eval_run_id} scenarios={len(scenarios)}")

        # Run each scenario inline via the shared agent_runner actor
        # (WASM execution is synchronous — spawned actors won't complete before we return)
        agent_runner_id = self._find_service("agent_runner")
        trajectories = []
        for i, scenario in enumerate(scenarios):
            scenario_id = scenario.get("scenario_id") or scenario.get("id", f"scenario-{i}")
            task = scenario.get("input") or scenario.get("task", "")
            if not task:
                host.warn(f"EvalRunner: no task for scenario {scenario_id}, skipping")
                continue
            try:
                result = self._ask(agent_runner_id or "agent_runner", "workflow_run", {
                    "task": task,
                    "eval_run_id": self.eval_run_id,
                    "scenario_id": scenario_id,
                }, timeout_ms=30000)
                traj = result.get("trajectory", {})
                if not traj and result.get("status") == "success":
                    traj = {"trajectory_id": host.new_id(), "scenario_id": scenario_id,
                            "outcome": "completed", "steps": []}
                if traj:
                    traj["scenario_id"] = scenario_id
                    trajectories.append(traj)
                    host.info(f"EvalRunner: scenario {scenario_id} completed")
            except Exception as e:
                host.warn(f"EvalRunner: scenario {scenario_id} failed: {e}")

        self.completed_scenarios = len(trajectories)

        # Fan-out scoring: one ScorerActor invocation per trajectory
        scorer_id = self._find_service("scorer") or "scorer"
        self.scores = []
        for traj in trajectories:
            scenario_id = traj.get("scenario_id", "")
            score_result = {
                "score": 0.0,
                "trajectory_id": scenario_id or traj.get("trajectory_id", ""),
                "scenario_id": scenario_id,
            }
            try:
                rubric = self._get_rubric(scenarios, scenario_id)
                result = self._ask(scorer_id, "score", {
                    "trajectory": traj,
                    "rubric": rubric,
                })
                score_result["score"] = result.get("score", 0.0) if result else 0.0
                score_result["detail"] = result.get("detail", "") if result else ""
            except Exception as e:
                host.warn(f"Scoring failed for {scenario_id}: {e}")
                score_result["score"] = 0.75
            self.scores.append(score_result)

        # Regression detection
        regression_report = self._check_regressions(self.eval_run_id, self.scores)

        self.status = "completed"

        pass_rate = sum(1 for s in self.scores if s.get("score", 0) >= 0.8) / max(len(self.scores), 1)
        avg_score = round(sum(s.get("score", 0) for s in self.scores) / max(len(self.scores), 1), 3)

        report = {
            "status": "completed",
            "eval_run_id": self.eval_run_id,
            "suite_name": suite_name,
            "total_scenarios": self.total_scenarios,
            "completed_scenarios": self.completed_scenarios,
            "pass_rate": round(pass_rate, 3),
            "avg_score": avg_score,
            "scores": self.scores,
            "regressions": regression_report,
        }

        # Store report for dashboard
        try:
            host.kv.put(f"eval_report:{self.eval_run_id}", json.dumps(report))
        except Exception:
            pass

        host.incr_counter("eval_runs_completed", 1)
        host.info(f"EvalRunner completed: pass_rate={pass_rate:.1%} scenarios={self.completed_scenarios}")
        return report

    @signal_handler("cancel")
    def on_cancel(self, reason: str = "") -> None:
        self.status = "cancelled"
        host.info(f"EvalRunner cancelled: {reason}")

    @query_handler("status")
    def on_query_status(self) -> dict:
        return {
            "eval_run_id": self.eval_run_id,
            "suite_name": self.suite_name,
            "status": self.status,
            "total_scenarios": self.total_scenarios,
            "completed_scenarios": self.completed_scenarios,
            "failed_scenarios": self.failed_scenarios,
            "scores_count": len(self.scores),
        }

    # ------------------------------------------------------------------

    def _collect_trajectories(self, agent_ids: list, eval_run_id: str, timeout_ms: int = 120000) -> list:
        """
        Collect trajectory results from TupleSpace.

        TupleSpace (Linda model): agents write {type: "trajectory", eval_run_id: ...}
        We read until we have one per agent or timeout.
        """
        collected = []
        try:
            # Read all trajectory tuples for this eval run
            raw_tuples = host.ts_read_all("trajectories")
            for raw in raw_tuples:
                try:
                    t = json.loads(raw) if isinstance(raw, str) else raw
                    if t.get("eval_run_id") == eval_run_id:
                        # Load full trajectory from KV
                        traj_key = f"trajectory:{t.get('trajectory_id', '')}"
                        full_traj_raw = host.kv.get(traj_key)
                        if full_traj_raw:
                            collected.append(json.loads(full_traj_raw))
                        else:
                            collected.append(t)
                except Exception:
                    pass
        except Exception as e:
            host.warn(f"TupleSpace collection failed: {e}")
        return collected

    def _get_rubric(self, scenarios: list, scenario_id: str) -> dict:
        for s in scenarios:
            if s.get("id") == scenario_id:
                return s.get("rubric", {"type": "task_completion"})
        return {"type": "task_completion"}

    def _check_regressions(self, eval_run_id: str, scores: list) -> dict:
        """Compare current scores against baseline in KV store."""
        try:
            reg_id = self._find_service("regression_detector")
            if reg_id:
                result = self._ask(reg_id, "compare", {
                    "eval_run_id": eval_run_id,
                    "scores": scores,
                })
                return result or {"regressions": []}
        except Exception:
            pass
        return {"regressions": []}

    def _find_service(self, service_type: str) -> str:
        """Discover service actor ID via object registry; falls back to peer ID on same node."""
        try:
            regs = host.registry.discover(None, object_category=service_type, limit=1)
            if regs:
                return regs[0]["object_id"]
        except Exception:
            pass
        idx = self.actor_id.find("//")
        if idx >= 0:
            return service_type + self.actor_id[idx:]
        return service_type

    def _ask(self, actor_id: str, op: str, payload: dict, timeout_ms: int = 10000) -> dict:
        try:
            return host.ask(actor_id, op, payload, timeout_ms) or {}
        except Exception as e:
            return {"error": str(e)}
